crates/wash-lib/src/build.rs

```
@@ -0,0 +1,205 @@
+    actor_config: ActorConfig,
+    language_config: LanguageConfig,
+    common_config: CommonConfig,
+    no_sign: bool,
+) -> Result<PathBuf> {
+    // Build actor based on language toolchain
+    let file_path = match language_config {
+        LanguageConfig::Rust(rust_config) => {
+            build_rust_actor(common_config.clone(), rust_config, actor_config.clone())
+        }
+        LanguageConfig::TinyGo(tinygo_config) => {
+            build_tinygo_actor(common_config.clone(), tinygo_config)
+        }
+    }?;
+
+    // Exit early if signing isn't desired
+    if no_sign {
+        Ok(file_path)
+    } else {
+        let source = file_path
+            .to_str()
+            .ok_or_else(|| anyhow!("Could not convert file path to string"))?
+            .to_string();
+
+        let destination = format!("build/{}_s.wasm", common_config.name);
+        let destination_file = PathBuf::from_str(&destination);
+
+        let sign_options = SignCommand {
+            source,
+            destination: Some(destination),
+            metadata: ActorMetadata {
+                name: common_config.name,
+                ver: Some(common_config.version.to_string()),
+                custom_caps: actor_config.claims,
+                call_alias: actor_config.call_alias,
+                ..Default::default()
+            },
+        };
+        sign_file(sign_options, OutputKind::Json)?;
+
+        Ok(destination_file?)
+    }
+}
+
+/// Builds a rust actor and returns the path to the file.
+fn build_rust_actor(
+    common_config: CommonConfig,
+    rust_config: RustConfig,
+    actor_config: ActorConfig,
+) -> Result<PathBuf> {
+    let mut command = match rust_config.cargo_path {
+        Some(path) => process::Command::new(path),
+        None => process::Command::new("cargo"),
+    };
+
+    let result = command.args(["build", "--release"]).status()?;
+
+    if !result.success() {
+        bail!("Compiling actor failed: {}", result.to_string())
```

Should a library `bail!`? Feels a bit extreme. Why not return an error and let the caller handle it?

---

That's what this does 😄 https://docs.rs/anyhow/1.0.66/anyhow/macro.bail.html super useful from `anyhow`

---

🤦 I totally thought this was essentially a prettier panic. TIL